Divs with border-radius are not repainted correctly when bounds change

vitorroriz opened the pull request:
#### 01b7c37f84bf7bb9cb4815f5cb06af7cabb7f798
<pre>
Divs with border-radius are not repainted correctly when bounds change
<a href="https://bugs.webkit.org/show_bug.cgi?id=116510">https://bugs.webkit.org/show_bug.cgi?id=116510</a>

Reviewed by Alan Bujtas.

* LayoutTests/imported/blink/fast/repaint/border-radius-repaint-2-expected.txt: Added.
* LayoutTests/imported/blink/fast/repaint/border-radius-repaint-2.html: Added.
* LayoutTests/platform/ios/fast/repaint/border-radius-repaint-2-expected.txt: Added.
* LayoutTests/platform/win/fast/repaint/border-radius-repaint-2-expected.txt: Added.
* LayoutTests/platform/gtk/fast/repaint/border-radius-repaint-2-expected.txt: Added.
* LayoutTests/platform/mac-bigsur-wk1/fast/repaint/border-radius-repaint-2-expected.txt: Added.
Adding new repaint reduced test and expected files

* Source/WebCore/platform/graphics/RoundedRect.h:
(WebCore::operator!=):
Adding operator!= for RoundedRect::Radii

* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::repaintAfterLayoutIfNeeded):
If bounds have changd and the element has border-radius, we should issue a full repaint
if border-radius for new bounds differs from old one.

Canonical link: <a href="https://commits.webkit.org/254041@main">https://commits.webkit.org/254041@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a0f2e864724cc6a62277d886583cc1b29b1c1cf4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87882 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31999 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18595 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97049 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/151976 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91849 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30344 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26377 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79969 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91793 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93495 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24488 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/74583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24469 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79441 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67308 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/28094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/28163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14421 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2841 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/30229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/74583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/30164 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33696 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/imported/blink/fast/repaint/border-radius-repaint-2.html:
```diff
@@ -0,0 +1,27 @@
+<!doctype html>
```
New tests should not go into an `imported` directory (unless that's `web-platform-tests` and you plan to export back to WPT).

vitorroriz replied:
Yes, this test is currently not on WPT so @brentfulgham suggested we could import it and then maybe export it to WPT as you mentioned. Should we create a new test?

vitorroriz replied:
I've decided to create a new similar test, that is a bit more reduced (we don't need the 2 extra divs) and that uses our repaint.js file for the repainting instead.

alanbaradlay commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1182,6 +1182,14 @@ bool RenderElement::repaintAfterLayoutIfNeeded(const RenderLayerModelObject* rep
+        auto oldRadius = style().getRoundedBorderFor(oldBounds).radii();
+        auto newRadius = style().getRoundedBorderFor(newBounds).radii();
+
+        fullRepaint = oldRadius != newRadius;
```
I am bit confused. if there's a geometry change (oldBounds != newBounds) aren't we supposed to issue a repaint on the enclosing geometry a few lines below?

vitorroriz replied:
Do you mean `mustRepaintBackgroundOrBorder(*this) && (newBounds != oldBounds || newOutlineBox != oldOutlineBox)` ?

It seems that `mustRepaintBackgroundOrBorder`  when verifying the border case just returns true   `if (renderer.style().hasBorder() && renderer.borderImageIsLoadedAndCanBeRendered())`


It doesn't check for differences on border-radius. Should we move the border-radius comparison to inside this function instead?

smfr commented on LayoutTests/fast/repaint/border-radius-repaint-2-expected.txt:
```diff
@@ -0,0 +1,7 @@
+layer at (0,0) size 800x600
```
This doesn't look like the correct output for a repaint test.

vitorroriz replied:
Sorry, I haven't noticed that repaint.js was deprecated by text-based-repaint.js.

I've updated it to use text-based and added a platform/iOS specific -expected file, since the rects logged are slightly different on iOS. Is this expected @smfr ? (I've tried not to use relative sizes)

vitorroriz committed:
Divs with border-radius are not repainted correctly when bounds change
https://bugs.webkit.org/show_bug.cgi?id=116510

Reviewed by Alan Bujtas.

* LayoutTests/imported/blink/fast/repaint/border-radius-repaint-2-expected.txt: Added.
* LayoutTests/imported/blink/fast/repaint/border-radius-repaint-2.html: Added.
* LayoutTests/platform/ios/fast/repaint/border-radius-repaint-2-expected.txt: Added.
* LayoutTests/platform/win/fast/repaint/border-radius-repaint-2-expected.txt: Added.
* LayoutTests/platform/gtk/fast/repaint/border-radius-repaint-2-expected.txt: Added.
* LayoutTests/platform/mac-bigsur-wk1/fast/repaint/border-radius-repaint-2-expected.txt: Added.
Adding new repaint reduced test and expected files

* Source/WebCore/platform/graphics/RoundedRect.h:
(WebCore::operator!=):
Adding operator!= for RoundedRect::Radii

* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::repaintAfterLayoutIfNeeded):
If bounds have changd and the element has border-radius, we should issue a full repaint
if border-radius for new bounds differs from old one.

Canonical link: https://commits.webkit.org/254041@main
